File: app/src/main/java/com/ebsoft/watchlist/ui/adapter/ListAdapter.java
package com.ebsoft.watchlist.ui.adapter;

import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by endre on 08/09/18.
 */

public class ListAdapter<E> extends RecyclerView.Adapter<ListViewHolder<E>>
        implements BindableAdapter<List<E>> {

    private final int mLayoutId;

    private final List<E> mDataSet;

    private ListItemClickListener<E> mItemClickListener;

    private ListItemRemoveListener<E> mItemRemoveListener;

    public ListAdapter(int layoutId) {
        this.mLayoutId = layoutId;
        this.mDataSet = new ArrayList<>();
    }

    public void setItemClickListener(ListItemClickListener<E> listener) {
        this.mItemClickListener = listener;
    }

    public void setItemRemoveListener(ListItemRemoveListener<E> listener) {
        this.mItemRemoveListener = listener;
    }

    @Override
    public void addItems(List<E> items) {
        mDataSet.addAll(items);
        notifyDataSetChanged();
    }

    @Override
    public void clearItems() {
        mDataSet.clear();
    }

    @Override
    public ListViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater =
                LayoutInflater.from(parent.getContext());
        ViewDataBinding binding = DataBindingUtil.inflate(
                layoutInflater, viewType, parent, false);
        return new ListViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ListViewHolder viewHolder, int position) {
        viewHolder.bind(mDataSet.get(position));
        viewHolder.setClickListener(getOnClickListenerForPosition(position));
        viewHolder.setRemoveClickListener(getItemRemoveListenerForPosition(position));
    }

    @Override
    public int getItemViewType(int position) {
        return mLayoutId;
    }

    @Override
    public int getItemCount() {
        return mDataSet.size();
    }

    private View.OnClickListener getOnClickListenerForPosition(int position) {
        return view -> {
            if (mItemClickListener != null) {
                mItemClickListener.onItemClick(mDataSet.get(position));
            }
        };
    }

    private View.OnClickListener getItemRemoveListenerForPosition(int position) {
        return view -> {
            if (mItemRemoveListener != null) {
                mItemRemoveListener.onRemove(mDataSet.get(position));
            }
        };
    }
}

File: app/src/main/java/com/ebsoft/watchlist/ui/adapter/ListAdapter.kt
package com.ebsoft.watchlist.ui.adapter

import android.databinding.DataBindingUtil
import android.databinding.ViewDataBinding
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import java.util.*

/**
 * Created by endre on 08/09/18.
 */

class ListAdapter<E>(private val mLayoutId: Int) : RecyclerView.Adapter<ListViewHolder<E>>(), BindableAdapter<List<E>> {

    private val mDataSet: MutableList<E>

    private var mItemClickListener: ListItemClickListener<E>? = null

    private var mItemRemoveListener: ListItemRemoveListener<E>? = null

    init {
        this.mDataSet = ArrayList()
    }

    fun setItemClickListener(listener: ListItemClickListener<E>) {
        this.mItemClickListener = listener
    }

    fun setItemRemoveListener(listener: ListItemRemoveListener<E>) {
        this.mItemRemoveListener = listener
    }

    override fun addItems(items: List<E>) {
        mDataSet.addAll(items)
        notifyDataSetChanged()
    }

    override fun clearItems() {
        mDataSet.clear()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ListViewHolder<E> {
        val layoutInflater = LayoutInflater.from(parent.context)
        val binding = DataBindingUtil.inflate<ViewDataBinding>(
                layoutInflater, viewType, parent, false)
        return ListViewHolder(binding)
    }

    override fun onBindViewHolder(viewHolder: ListViewHolder<E>, position: Int) {
        viewHolder.bind(mDataSet[position])
        viewHolder.setClickListener(getOnClickListenerForPosition(position))
        viewHolder.setRemoveClickListener(getItemRemoveListenerForPosition(position))
    }

    override fun getItemViewType(position: Int): Int {
        return mLayoutId
    }

    override fun getItemCount(): Int {
        return mDataSet.size
    }

    private fun getOnClickListenerForPosition(position: Int): View.OnClickListener {
        return View.OnClickListener {
            if (mItemClickListener != null) {
                mItemClickListener!!.onItemClick(mDataSet[position])
            }
        }
    }

    private fun getItemRemoveListenerForPosition(position: Int): View.OnClickListener {
        return View.OnClickListener {
            if (mItemRemoveListener != null) {
                mItemRemoveListener!!.onRemove(mDataSet[position])
            }
        }
    }
}
